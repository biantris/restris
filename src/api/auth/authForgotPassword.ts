import UserModel from "../../modules/user/UserModel";

export const authForgotPassword = async (ctx) => {
  const user = await UserModel.findOne({
    email: ctx.query.user.trim(),
  });

  ctx.status = 200;

  if (!user) {
    ctx.body = {
      message: "User not found",
    };
    return;
  }

  ctx.body = {
    user,
    password: 123456, //change to user.password
  };
  return;
};
