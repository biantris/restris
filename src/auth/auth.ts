import { Base64String, unbase64 } from "./base64";
import { getObjectId } from "../../test";

import User from "../modules/user/UserModel";
import jwt_decode from "jwt-decode";

export type Token = {
  userId: string | null;
};

export type TokenDecoded = {
  user: string;
}

export const getToken = (authorization: Base64String): Token => {
  const tokenDecoded: TokenDecoded = jwt_decode(unbase64(authorization));
  const userId = tokenDecoded.user;
  
  if (!userId) {
    return {
      userId: null,
    };
  }

  return {
    userId: userId,
  };
};

export const getUser = async (token: string) => {
  const { userId } = getToken(token);

  if (!userId) {
    return null;
  }

  const user = await User.findOne({
    _id: getObjectId(userId),
    removedAt: null,
  });

  return user;
};

export const auth = async (ctx, next) => {
  const { authorization } = ctx.header;

  if (!authorization) {
    ctx.status = 401;
    ctx.body = {
      message: "Unauthorized",
    };
    return;
  }

  const user = await getUser(authorization);

  if (!user) {
    ctx.status = 401;
    ctx.body = {
      message: "Unauthorized",
    };
    return;
  }

  ctx.user = user;

  await next();
};
