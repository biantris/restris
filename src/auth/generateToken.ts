import jwt from "jsonwebtoken";
import { IUser } from "../../src/modules/user/UserModel";

import { config } from "../config";

export const generateToken = (user: IUser) => {
  return `JWT ${jwt.sign({ user: user._id }, config.JWT_SECRET)}`;
};
