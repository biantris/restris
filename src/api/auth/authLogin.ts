import User from "../../modules/user/UserModel";
import { generateToken } from "../../auth/generateToken";
import { base64 } from "../../auth/base64";

export const authLogin = async (ctx) => {
  const { email, password } = ctx.request.body;

  if (!email || !password) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  const user = await User.findOne({
    email,
    removeAt: null,
  });

  if (!user) {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  let correctPassword;

  try {
    correctPassword = user.authenticate(password);
  } catch {
    ctx.status = 401;
    ctx.body = {
      message: "Email or password incorrect",
    };
    return;
  }

  if(!correctPassword) {
    ctx.status = 401;
    ctx.body = {
        message: "Email or password incorrect",
    }
    return;
  }

  ctx.status = 200;
  ctx.body = {
    message: "User authenticated with success",
    token: base64(generateToken(user)),
    user
  }
};
