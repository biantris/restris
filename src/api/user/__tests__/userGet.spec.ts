import {
  clearDbAndRestartCounters,
  connectMongoose,
  disconnectMongoose,
  createGetApiCall,
  sanitizeTestObject,
} from "../../../../test";
import { createUser } from "../../../modules/user/fixtures/createUser";
import { base64 } from "../../../auth/base64";
import { generateToken } from "../../../auth/generateToken";

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const getUrl = (id: string) => `/api/user/${id}`;

it("should return error if authorization header does not exist", async () => {
  const response = await createGetApiCall({
    url: getUrl("5c42132aa591a2001ad46264"),
  });

  expect(response.status).toBe(401);
  expect(response.body.message).toBe("Unauthorized");
  expect(response.body).toMatchSnapshot();
});

it("should return error if user it was not found", async () => {
  const admin = await createUser({});
  const authorization = base64(generateToken(admin));

  const response = await createGetApiCall({
    url: getUrl("5c42132aa591a2001ad46264"),
    authorization,
  });

  expect(response.status).toBe(400);
  expect(response.body.message).toBe("User not found");
  expect(response.body).toMatchSnapshot();
});

it("should return user by id", async () => {
  const admin = await createUser({});
  const authorization = base64(generateToken(admin));
  const user = await createUser({});
  const response = await createGetApiCall({
    url: getUrl(user._id),
    authorization,
  })

  expect(response.status).toBe(200);
  expect(response.body.user.name).toBe(user.name);
  expect(response.body.user.email).toBe(user.email);
  expect(sanitizeTestObject(response.body)).toMatchSnapshot();
});
