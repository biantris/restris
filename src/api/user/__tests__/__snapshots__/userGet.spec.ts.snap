// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return error if authorization header does not exist 1`] = `
Object {
  "message": "Unauthorized",
}
`;

exports[`should return error if user it was not found 1`] = `
Object {
  "message": "User not found",
}
`;

exports[`should return user by id 1`] = `
Object {
  "user": Object {
    "_id": "ObjectId",
    "email": "user2@test.com",
    "name": "User 2",
  },
}
`;
